Fix #3491 and #3479: App layer mechanism to resume exploration

MaskedCarrot opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fixes #3491 and #3479 

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### Espresso Tests
#### StateFragmentTest
![Screenshot from 2021-08-21 04-19-15](https://user-images.githubusercontent.com/53645584/130322381-e52a67c9-562e-4c88-861f-1b84dbd2d2ab.png)

#### ExplorationActivityTest
![Screenshot from 2021-08-21 04-47-49](https://user-images.githubusercontent.com/53645584/130322392-7755faa8-c405-4168-b903-7f203ddc7703.png)

#### TopicLessonsFragmentTest
![Screenshot from 2021-08-21 18-37-18](https://user-images.githubusercontent.com/53645584/130323032-1a0a767b-86b7-4166-9844-c403e4fab905.png)

Commit:
moved hint handler to domain layer

yashraj-01 committed:
fixed app layer espresso tests

yashraj-01 committed:
fix app layer robolectric tests

yashraj-01 committed:
fixed domain layer tests

yashraj-01 committed:
Added annotations to test exemptions

yashraj-01 committed:
proto lint fix

Commit:
fixed hint handler for training sessions

Commit:
nit and removed test excemptions

Commit:
added hint tests for config change

Commit:
fixed test file exemptions

Commit:
fixed failing test

yashraj-01 committed:
made HintHandler injectable

yashraj-01 committed:
Merge remote-tracking branch 'upstream/move-hint-handler-to-domain' into move-hint-handler-to-domain

yashraj-01 committed:
fixed ktlint error

yashraj-01 committed:
Added tests for hint handler

MaskedCarrot committed:
Merge branch 'develop' into move-hint-handler-to-domain

Commit:
nits

Commit:
Merge remote-tracking branch 'origin/move-hint-handler-to-domain' into move-hint-handler-to-domain

Commit:
fixed failing build

Commit:
fixed failing build

Commit:
fix build

Commit:
fixed imports

MaskedCarrot committed:
Merge branch 'develop' into move-hint-handler-to-domain

Commit:
completed mechansim to resume exploration

Commit:
nits and improved testing

Commit:
updated exploration.proto

Commit:
removed progress controller from kdoc exemptions

Commit:
Merge branch 'move-hint-handler-to-domain' into resume-exploration-in-domain-module

Commit:
added kdoc

Commit:
fixed failing build

Commit:
added resume lesson activity

Commit:
nit fix

Commit:
Merge remote-tracking branch 'origin/develop' into move-hint-handler-to-domain

Commit:
fix failing test

Commit:
Merge branch 'move-hint-handler-to-domain' into resume-exploration-in-domain-module

Commit:
Merge branch 'resume-exploration-in-domain-module' into add-resume-lesson-activity

Commit:
finsihed mech to resume exploration

Commit:
fixed failing build

Commit:
Revert "fixed failing build"

This reverts commit 6ba0249be01a4a8f91ecdadb702277b3c086edbd.

Commit:
fixed failing build

Commit:
Merge branch 'add-resume-lesson-activity' into app-layer-mech-to-resume-checkpoints

MaskedCarrot commented:
closing this PR because it contains a lot of unnecessary changes, the new PR is #3672

Commit:
fix failing build

Commit:
fixed tests

MaskedCarrot committed:
updated layout files

MaskedCarrot commented on app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt:
```diff
@@ -221,17 +226,96 @@ class RecentlyPlayedFragmentPresenter @Inject constructor(
+        ProfileId.getDefaultInstance(),
+        promotedStory.explorationId
+      ).toLiveData().observe(
+        fragment,
+        Observer {
+          if (it.isSuccess()) {
+            startOrResumeExploration(
+              internalProfileId,
+              promotedStory.topicId,
+              promotedStory.storyId,
+              promotedStory.explorationId,
+              shouldSavePartialProgress,
+              canExplorationBeResumed = it.getOrThrow(),
+              backflowScreen = null
+            )
+          } else if (it.isFailure()) {
+            startOrResumeExploration(
+              internalProfileId,
+              promotedStory.topicId,
+              promotedStory.storyId,
+              promotedStory.explorationId,
+              shouldSavePartialProgress,
+              canExplorationBeResumed = false,
+              backflowScreen = null
+            )
+          }
+        }
+      )
+    } else {
+      startOrResumeExploration(
+        internalProfileId,
+        promotedStory.topicId,
+        promotedStory.storyId,
+        promotedStory.explorationId,
+        shouldSavePartialProgress,
+        canExplorationBeResumed = false,
+        backflowScreen = null
+      )
+    }
   }
```
There are two erros due to this code:

1. If the exploration version doesn't match, both resume lesson and exploration activity are opened one after the other, when only one should have been opened.


I think the reason because exploration Activity and ResumeLessonActivity is starting are starting (one after the other) if the exploration doesn't match is since when the exploration versions don't match we start first call play exploration (that loads exploration and then saved checkpoint) and then open exploration activity, in the mean time version of exploration is updated (because play exploration saves a new checkpoint), the observer is notified and this time resume exploration activity is opened.




2. This error will surely not occur in real-world scenarios so I am not sure that if we should take care of this: if the exploration version is updated while the user is in the exploration (cannot occur in real-world scenarios), when the user clicks back to go back to the topic lessons fragment, upon returning to the topic lessons fragment they are automatically taken back to the exploration.

I am not sure why this is happening, but I think it might be because the version is now updated so the observer (that matched the version in the topic lessons fragment) starts the exploration again.

BenHenning replied:
My suggestion elsewhere to change the flow to navigate based on the state of the checkpoint DataProvider might solve this (presuming we're doing that in ExplorationFragment) since we wouldn't be passing in an actual checkpoint to initiate the exploration if it's incompatible. Merging the two DataProviders together should avoid the race condition described here.

MaskedCarrot replied:
as per chats in hangouts, used removeObserver.

MaskedCarrot reviewed (commented):
@BenHenning PTAL

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointController.kt:
```diff
@@ -228,6 +235,31 @@ class ExplorationCheckpointController @Inject constructor(
+        } else
+          AsyncResult.failed(
+            ExplorationCheckpointNotFoundException(
+              "Checkpoint with the explorationId $explorationId was not found " +
+                "for profileId ${profileId.internalId}."
```
```suggestion
              "Checkpoint with the exploration $explorationId was not found " +
                "for profile ${profileId.internalId}."
```

MaskedCarrot replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointController.kt:
```diff
@@ -228,6 +235,31 @@ class ExplorationCheckpointController @Inject constructor(
     }
   }
 
+  fun isSavedCheckpointCompatibleWithExploration(
```
Needs KDoc & tests.

MaskedCarrot replied:
done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointController.kt:
```diff
@@ -228,6 +235,31 @@ class ExplorationCheckpointController @Inject constructor(
+        CHECK_IS_EXPLORATION_CHECKPOINT_COMPATIBLE_WITH_EXPLORATION_DATA_PROVIDER_ID
+      ) { explorationCheckpointDatabase ->
+        val checkpoint = explorationCheckpointDatabase.explorationCheckpointMap[explorationId]
+        val chapterDataJsonObject = jsonAssetRetriever.loadJsonFromAsset("$explorationId.json")
```
We should be using ExplorationRetriever here, instead, so that this can work correctly when loading from proto (we currently support loading topic data from both protos and json).

MaskedCarrot replied:
done

BenHenning commented on app/src/main/java/org/oppia/android/app/story/storyitemviewmodel/StoryChapterSummaryViewModel.kt:
```diff
@@ -33,16 +38,100 @@ class StoryChapterSummaryViewModel(
+    }
+  }
+
+  private fun startOrResumeExploration(
+    internalProfileId: Int,
+    topicId: String,
+    storyId: String,
+    explorationId: String,
+    shouldSavePartialProgress: Boolean,
+    canExplorationBeResumed: Boolean,
+    backflowScreen: Int?
+  ) {
+    if (canExplorationBeResumed) {
+      explorationSelectionListener.selectExploration(
+        internalProfileId,
+        topicId,
+        storyId,
+        explorationId,
+        canExplorationBeResumed,
+        shouldSavePartialProgress,
+        backflowScreen
+      )
+    } else {
+      playExploration(
+        canExplorationBeResumed,
+        shouldSavePartialProgress,
+        backflowScreen
+      )
+    }
+  }
```
If we need to check at all locations whether there is compatibility, and in that case we just abandon the checkpoint & start over, why not move the version check into the DataProvider that provides the checkpoint, return a default, and short-circuit to ExplorationActivity if no checkpoint was found?

MaskedCarrot replied:
As per our chat, used `removeObserver`

BenHenning commented on domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt:
```diff
@@ -641,6 +647,9 @@ class TopicListController @Inject constructor(
         storyRecord.chaptersList.firstOrNull()?.let {
           nextChapterName = it.title
           explorationId = it.explorationId
+          // ChapterPlayState will be NOT_STARTED because this function only recommends first story
+          // of un-started topics.
```
```suggestion
          // ChapterPlayState will be NOT_STARTED because this function only recommends the first
          // story of un-started topics.
```

MaskedCarrot replied:
done

BenHenning reviewed (changes requested):
Thanks @MaskedCarrot. Took an initial pass (primarily on my codeowner files).

MaskedCarrot committed:
moved timer to domain

MaskedCarrot committed:
fixed build and nits

MaskedCarrot committed:
merge base branch

MaskedCarrot committed:
moved functions to ExplorationProgress.kt

MaskedCarrot committed:
Merge branch 'resume-exploration-in-domain-module' into add-resume-lesson-activity

MaskedCarrot committed:
nit

MaskedCarrot committed:
nit

MaskedCarrot committed:
Merge branch 'move-hint-handler-to-domain' into resume-exploration-in-domain-module

MaskedCarrot committed:
Merge branch 'resume-exploration-in-domain-module' into add-resume-lesson-activity

MaskedCarrot committed:
added tests

MaskedCarrot committed:
updated exemptions

MaskedCarrot committed:
added listener back to test exemptions

MaskedCarrot committed:
Merge branch 'move-hint-handler-to-domain' into resume-exploration-in-domain-module

MaskedCarrot committed:
Merge branch 'resume-exploration-in-domain-module' into add-resume-lesson-activity

MaskedCarrot committed:
nit

MaskedCarrot committed:
Merge branch 'add-resume-lesson-activity' into app-layer-mech-to-resume-checkpoints

MaskedCarrot committed:
nit

MaskedCarrot committed:
updated proto

MaskedCarrot committed:
nit fixes and added more tests

MaskedCarrot committed:
lint fix

MaskedCarrot committed:
Merge branch 'move-hint-handler-to-domain' into resume-exploration-in-domain-module

MaskedCarrot committed:
nit fix

MaskedCarrot committed:
Merge branch 'resume-exploration-in-domain-module' into add-resume-lesson-activity

MaskedCarrot committed:
nit fix

MaskedCarrot committed:
removed unused fun

MaskedCarrot committed:
Merge branch 'resume-exploration-in-domain-module' into add-resume-lesson-activity

MaskedCarrot committed:
updated xml

BenHenning committed:
First round: make HintHandler independent.

This also brings HintHandler into an interface + factory pattern. This
isn't the final design since I think we can largely simplify the way
hints work; that'll be my next pass.

This breaks questions & HintHandlerTest; those will require further work
later.

BenHenning committed:
Simplify HelpIndex in proto.

Move HelpIndex to PendingState to avoid the entire domain case of
handling CompletedState.

MaskedCarrot committed:
updated mech to start exploration

MaskedCarrot committed:
Merge branch 'add-resume-lesson-activity' into app-layer-mech-to-resume-checkpoints

BenHenning committed:
Simplify hints & solutions.

This removes the per-Hint/Solution tracking & completely leans on
HelpIndex for proper hints & solution tracking both in the domain & UI
layers.

Fixes a bunch of other stuff, too, including
QuestionAssessmentProgressController tests.

MaskedCarrot committed:
revert files

BenHenning committed:
Clean up dead code paths & improve handler API.

MaskedCarrot committed:
updated xml and nits

MaskedCarrot committed:
Merge branch 'add-resume-lesson-activity' into app-layer-mech-to-resume-checkpoints

MaskedCarrot committed:
added tests

yashraj-01 committed:
fixed test modules and lint

yashraj-01 committed:
renamed HintHandlerTest to HintHandlerImplTest

MaskedCarrot committed:
updated tests

MaskedCarrot committed:
revert file

MaskedCarrot committed:
fixed failing tests

BenHenning committed:
Add tests for HintHandler.

This introduces some new explorations for making testing HintHandlerImpl
easier.

MaskedCarrot committed:
updated xml

BenHenning committed:
Add remaining tests/exemptions for new files.

BenHenning committed:
Lint fixes.

MaskedCarrot committed:
nit

BenHenning committed:
Merge branch 'develop' into move-hint-handler-to-domain

Conflicts:
	app/src/main/java/org/oppia/android/app/application/ApplicationComponent.kt
	app/src/sharedTest/java/org/oppia/android/app/home/TopicSummaryViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/home/WelcomeViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/home/promotedlist/PromotedStoryListViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/home/promotedlist/PromotedStoryViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/parser/CustomBulletSpanTest.kt
	app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt
	app/src/sharedTest/java/org/oppia/android/app/splash/SplashActivityTest.kt
	app/src/test/java/org/oppia/android/app/home/HomeActivityLocalTest.kt
	app/src/test/java/org/oppia/android/app/player/exploration/ExplorationActivityLocalTest.kt
	app/src/test/java/org/oppia/android/app/profile/ProfileChooserFragmentLocalTest.kt
	app/src/test/java/org/oppia/android/app/testing/CompletedStoryListSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/HomeSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/OngoingTopicListSpanTest.kt
app/src/test/java/org/oppia/android/app/testing/ProfileChooserSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/ProfileProgressSpanCount.kt
	app/src/test/java/org/oppia/android/app/testing/RecentlyPlayedSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/TopicRevisionSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/player/split/PlayerSplitScreenTesting.kt
	app/src/test/java/org/oppia/android/app/testing/player/state/StateFragmentAccessibilityTest.kt
	app/src/test/java/org/oppia/android/app/topic/revisioncard/RevisionCardActivityLocalTest.kt
	domain/src/test/java/org/oppia/android/domain/exploration/ExplorationDataControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/question/QuestionAssessmentProgressControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/question/QuestionTrainingControllerTest.kt

BenHenning committed:
Post-merge fixes (including lint fixes).

BenHenning committed:
Merge branch 'develop' into move-hint-handler-to-domain

Conflicts:
	third_party/maven_install.json

BenHenning committed:
Post-merge maven_install fix.

MaskedCarrot committed:
Merge branch 'develop' into move-hint-handler-to-domain

BenHenning commented:
Unless I'm missing something, isn't this still using 2 DataProviders @MaskedCarrot? Per our discussion, my understanding is that we should use just 1 to simplify the dataflow and reduce potential inconsistencies (though the one regarding the observer is unavoidable without using removeObserver as you mentioned).

MaskedCarrot committed:
Revert "Merge branch 'develop' into move-hint-handler-to-domain"

This reverts commit e2fea908a65af8c11b5ed135f59cec38e807f251, reversing
changes made to 6659858f0b85a0bf0c51f23af600769eac64b938.

BenHenning committed:
Post-merge Gradle-discovered fixes.

BenHenning committed:
Revert "Revert "Merge branch 'develop' into move-hint-handler-to-domain""

This reverts commit b1622c00d33f584ae727f0c9c4434ad773bc80a2.

BenHenning committed:
Merge branch 'attempt-to-fix-move-hint-handler-to-domain' into move-hint-handler-to-domain

Conflicts:
	app/src/sharedTest/java/org/oppia/android/app/home/TopicSummaryViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/home/WelcomeViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/home/promotedlist/PromotedStoryListViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/home/promotedlist/PromotedStoryViewModelTest.kt
	app/src/sharedTest/java/org/oppia/android/app/parser/CustomBulletSpanTest.kt
	app/src/sharedTest/java/org/oppia/android/app/player/state/StateFragmentTest.kt
	app/src/sharedTest/java/org/oppia/android/app/splash/SplashActivityTest.kt
	app/src/test/java/org/oppia/android/app/home/HomeActivityLocalTest.kt
	app/src/test/java/org/oppia/android/app/player/exploration/ExplorationActivityLocalTest.kt
	app/src/test/java/org/oppia/android/app/profile/ProfileChooserFragmentLocalTest.kt
	app/src/test/java/org/oppia/android/app/testing/CompletedStoryListSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/HomeSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/OngoingTopicListSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/ProfileChooserSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/ProfileProgressSpanCount.kt
	app/src/test/java/org/oppia/android/app/testing/RecentlyPlayedSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/TopicRevisionSpanTest.kt
	app/src/test/java/org/oppia/android/app/testing/player/split/PlayerSplitScreenTesting.kt
	app/src/test/java/org/oppia/android/app/testing/player/state/StateFragmentAccessibilityTest.kt
	app/src/test/java/org/oppia/android/app/topic/revisioncard/RevisionCardActivityLocalTest.kt
	domain/src/test/java/org/oppia/android/domain/exploration/ExplorationDataControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/exploration/ExplorationProgressControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/question/QuestionAssessmentProgressControllerTest.kt
	domain/src/test/java/org/oppia/android/domain/question/QuestionTrainingControllerTest.kt

BenHenning committed:
Additional post-merge fixes.

MaskedCarrot committed:
Merge branch 'move-hint-handler-to-domain' into resume-exploration-in-domain-module

MaskedCarrot committed:
updated appraoch to resume hints to work with the new hint handler.

MaskedCarrot committed:
nit fix

MaskedCarrot committed:
Merge branch 'resume-exploration-in-domain-module' into add-resume-lesson-activity

MaskedCarrot committed:
updated fragment component

MaskedCarrot committed:
updated topic controller

MaskedCarrot committed:
Merge branch 'add-resume-lesson-activity' into app-layer-mech-to-resume-checkpoints

MaskedCarrot committed:
updated xml and tests

MaskedCarrot committed:
Merge branch 'add-resume-lesson-activity' into app-layer-mech-to-resume-checkpoints

MaskedCarrot committed:
WIP update approach to use single data provider

MaskedCarrot committed:
completed appraoch

MaskedCarrot committed:
nit

MaskedCarrot committed:
Merge branch 'add-resume-lesson-activity' into app-layer-mech-to-resume-checkpoints

MaskedCarrot committed:
fix failing build

MaskedCarrot committed:
nit

MaskedCarrot committed:
nit

MaskedCarrot committed:
nit fix

MaskedCarrot committed:
Merge branch 'add-resume-lesson-activity' into app-layer-mech-to-resume-checkpoints

MaskedCarrot committed:
fixed clock

MaskedCarrot committed:
nit

BenHenning commented on domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt:
```diff
@@ -641,6 +647,9 @@ class TopicListController @Inject constructor(
         storyRecord.chaptersList.firstOrNull()?.let {
           nextChapterName = it.title
           explorationId = it.explorationId
+          // ChapterPlayState will be NOT_STARTED because this function only recommends the first
+          // story of un-started topics.
+          chapterPlayState = ChapterPlayState.NOT_STARTED
```
Should be moved outside the ``let`` block since it's unrelated.

MaskedCarrot replied:
moved outside `let`

BenHenning commented on domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt:
```diff
@@ -690,7 +700,8 @@ class TopicListController @Inject constructor(
     explorationId: String?,
-    isTopicConsideredCompleted: Boolean
+    isTopicConsideredCompleted: Boolean,
+    chapterProgress: ChapterProgress?
```
Can this name be more specific? Maybe ``nextChapterProgress``?

MaskedCarrot replied:
updated name.

BenHenning commented on domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationStorageModuleTest.kt:
```diff
@@ -93,6 +96,15 @@ class ExplorationStorageModuleTest {
     @Provides
     fun provideCacheAssetsLocally(): Boolean = false
+
+    @Provides
+    @TopicListToCache
+    fun provideTopicListToCache(): List<String> = listOf()
+
+    @Provides
+    @LoadLessonProtosFromAssets
+    fun provideLoadLessonProtosFromAssets(testEnvironmentConfig: TestEnvironmentConfig): Boolean =
+      testEnvironmentConfig.isUsingBazel()
```
Out of curiosity, why are these needed now? I wouldn't expect these to be needed for a test suite that's just verifying the module providers.

MaskedCarrot replied:
there are not needed here, removed them.

MaskedCarrot replied:
Added these back. Removing this is causing the tests to fail because we are injecting explorationcheckpointController here, which uses these dependencies.

BenHenning commented on testing/src/main/java/org/oppia/android/testing/lightweightcheckpointing/ExplorationCheckpointTestHelper.kt:
```diff
@@ -13,23 +13,37 @@ import org.mockito.MockitoAnnotations
-const val FAKE_EXPLORATION_TITLE_1 = "Fake exploration title_1"
-const val FAKE_EXPLORATION_ID_2 = "fake_exploration_id_2"
-const val FAKE_EXPLORATION_TITLE_2 = "Fake exploration title_2"
+const val FRACTIONS_EXPLORATION_0_TITLE = "What is a Fraction?"
```
Please add KDocs for all currently undocumented non-private members in this class & remove the exemption from scripts/assets/kdoc_validity_exemptions.textproto.

MaskedCarrot replied:
done

BenHenning commented on testing/src/main/java/org/oppia/android/testing/lightweightcheckpointing/ExplorationCheckpointTestHelper.kt:
```diff
@@ -49,36 +63,126 @@ class ExplorationCheckpointTestHelper @Inject constructor(
+   * Updates the saved checkpoint for Fractions, story 0, exploration 0. For this function to work
+   * correctly it should be called after [saveCheckpointForFractionsStory0Exploration0].
+   *
+   * @param profileId the profileID for which the checkpoint has to be saved
```
```suggestion
   * @param profileId the profile ID for which the checkpoint has to be saved
```

Ditto elsewhere in the PR.

MaskedCarrot replied:
done

BenHenning commented on testing/src/test/java/org/oppia/android/testing/lightweightcheckpointing/ExplorationCheckpointTestHelperTest.kt:
```diff
@@ -79,20 +89,77 @@ class ExplorationCheckpointTestHelperTest {
+    assertThat(explorationCheckpointCaptor.value.getOrThrow().explorationTitle)
+      .isEqualTo(FRACTIONS_EXPLORATION_0_TITLE)
+    assertThat(explorationCheckpointCaptor.value.getOrThrow().pendingStateName)
+      .isEqualTo(FRACTIONS_STORY_0_EXPLORATION_0_SECOND_STATE_NAME)
```
Could you also add a verification line for state name in the previous test? Otherwise it seems a bit like update*() could do nothing & both tests would still pass (it's not obvious what the "difference" state being verified here is).

Ditto for the other similar tests here.

MaskedCarrot replied:
done

BenHenning commented on testing/src/test/java/org/oppia/android/testing/lightweightcheckpointing/ExplorationCheckpointTestHelperTest.kt:
```diff
@@ -79,20 +89,77 @@ class ExplorationCheckpointTestHelperTest {
+  fun testSaveCheckpointForFractionsStory0Exploration0_checkCheckpointIsSaved() {
+    explorationCheckpointTestHelper.saveCheckpointForFractionsStory0Exploration0(
+      profileId = profileId,
+      version = FRACTIONS_STORY_0_EXPLORATION_0_CORRECT_VERSION
```
Here & elsewhere: should we also be verifying that the version is used correctly (i.e. saved in the checkpoint)?

Suggest also adding a test case for a different version being used.

MaskedCarrot replied:
we can only retrieve checkpoints if the version saved in the checkpoint matches with the current exploration version. Otherwise, a failure result is returned.

BenHenning commented on testing/src/test/java/org/oppia/android/testing/lightweightcheckpointing/ExplorationCheckpointTestHelperTest.kt:
```diff
@@ -102,43 +169,90 @@ class ExplorationCheckpointTestHelperTest {
+    assertThat(explorationCheckpointCaptor.value.getOrThrow().explorationTitle)
+      .isEqualTo(FRACTIONS_EXPLORATION_0_TITLE)
+    assertThat(explorationCheckpointCaptor.value.getOrThrow().pendingStateName)
+      .isEqualTo(FRACTIONS_STORY_0_EXPLORATION_0_SECOND_STATE_NAME)
```
This & above seem odd--why does a ratios checkpoint have fractions values in it?

MaskedCarrot replied:
updated.

BenHenning commented on domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointControllerTest.kt:
```diff
@@ -252,17 +295,45 @@ class ExplorationCheckpointControllerTest {
+  }
+
+  @Test
+  fun testController_saveInCompatibleCheckpoint_checkpointCheckpointIsNotCompatible() {
```
```suggestion
  fun testController_saveIncompatibleCheckpoint_checkpointCheckpointIsNotCompatible() {
```

BenHenning replied:
Also, typo in name? 'checkpointCheckpoint'

MaskedCarrot replied:
updated test name

BenHenning commented on domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointControllerTest.kt:
```diff
@@ -252,17 +295,45 @@ class ExplorationCheckpointControllerTest {
+      firstTestProfile,
+      FRACTIONS_EXPLORATION_ID_0
+    ).toLiveData().observeForever(mockExplorationCheckpointObserver)
+    verifyMockObserverIsFailure(mockExplorationCheckpointObserver, explorationCheckpointCaptor)
```
For this & all other tests: I suggest verifying the specific failure that's occurring to make sure that it's the failure you expect it to be.

MaskedCarrot replied:
done

BenHenning replied:
Thanks--I think that this really improved the clarity of the tests!

BenHenning commented on domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationCheckpointControllerTest.kt:
```diff
@@ -252,17 +295,45 @@ class ExplorationCheckpointControllerTest {
+  fun testController_saveInCompatibleCheckpoint_checkpointCheckpointIsNotCompatible() {
+    explorationCheckpointTestHelper.saveCheckpointForFractionsStory0Exploration0(
+      profileId = firstTestProfile,
+      version = FRACTIONS_STORY_0_EXPLORATION_0_INCORRECT_VERSION
```
Suggest instead: ``FRACTIONS_STORY_0_EXPLORATION_0_OLD_VERSION`` and ``FRACTIONS_STORY_0_EXPLORATION_0_CURRENT_VERSION`` names to be more specific.

MaskedCarrot replied:
done

BenHenning commented on app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt:
```diff
@@ -221,17 +224,71 @@ class RecentlyPlayedFragmentPresenter @Inject constructor(
+      explorationCheckpointLiveData.observe(
+        fragment,
+        object : Observer<AsyncResult<ExplorationCheckpoint>> {
+          override fun onChanged(it: AsyncResult<ExplorationCheckpoint>?) {
```
Is the null property actually necessary? I don't think we do this in other observers.

MaskedCarrot replied:
removed the null property

BenHenning reviewed (changes requested):
Thanks @MaskedCarrot. Took a full review per my codeowners and left some comments--PTAL. Also, please fix the failing CI checks.

Overall, I think the mechanism looks good.

MaskedCarrot committed:
Merge remote-tracking branch 'origin/develop' into resume-exploration-in-domain-module

MaskedCarrot committed:
fixed merge

MaskedCarrot committed:
test and lint fix

MaskedCarrot committed:
updated

MaskedCarrot committed:
Merge branch 'resume-exploration-in-domain-module' into add-resume-lesson-activity

MaskedCarrot committed:
nit fix

MaskedCarrot committed:
lint fix

MaskedCarrot committed:
Merge branch 'add-resume-lesson-activity' into app-layer-mech-to-resume-checkpoints

MaskedCarrot committed:
fix failing tests

MaskedCarrot committed:
Merge branch 'add-resume-lesson-activity' into app-layer-mech-to-resume-checkpoints

MaskedCarrot committed:
nit

MaskedCarrot committed:
fixed tests

MaskedCarrot committed:
lint fix

MaskedCarrot committed:
Merge branch 'resume-exploration-in-domain-module' into add-resume-lesson-activity

MaskedCarrot committed:
Merge branch 'add-resume-lesson-activity' into app-layer-mech-to-resume-checkpoints

MaskedCarrot committed:
fixed failing test

MaskedCarrot committed:
removed explorationCheckpointTestHelper from Kdoc exemptions

vinitamurthi reviewed (approved):
LGTM on codeowner files!

aggarwalpulkit596 commented on app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt:
```diff
@@ -221,17 +224,69 @@ class RecentlyPlayedFragmentPresenter @Inject constructor(
+        object : Observer<AsyncResult<ExplorationCheckpoint>> {
+          override fun onChanged(it: AsyncResult<ExplorationCheckpoint>) {
+            if (it.isSuccess()) {
+              explorationCheckpointLiveData.removeObserver(this)
```
Shouldn't we call it outside this if and else condition instead?

MaskedCarrot replied:
I think moving this outside the if-else will make it trigger when the async result becomes pending

anandwana001 commented on app/src/sharedTest/java/org/oppia/android/app/testing/TopicTestActivityForStoryTest.kt:
```diff
@@ -89,6 +90,9 @@ class TopicTestActivityForStoryTest {
   lateinit var testCoroutineDispatchers: TestCoroutineDispatchers
 
+  @Inject
+  lateinit var explorationRetriever: ExplorationRetriever
```
Do we need this?

MaskedCarrot replied:
remvoed

anandwana001 commented on domain/src/test/java/org/oppia/android/domain/exploration/lightweightcheckpointing/ExplorationStorageModuleTest.kt:
```diff
@@ -93,6 +95,11 @@ class ExplorationStorageModuleTest {
+
+    @Provides
+    @LoadLessonProtosFromAssets
+    fun provideLoadLessonProtosFromAssets(testEnvironmentConfig: TestEnvironmentConfig): Boolean =
```
Are we using this?

MaskedCarrot replied:
yes, we inject explorationCheckpointController here and we use these there.

BenHenning replied:
We also generally always want to use the switch logic when possible to ensure tests are verifying both the JSON & proto pipelines (in Gradle & Bazel, respectively).

anandwana001 reviewed (changes requested):
nit thoughts.

Also, add a screenshot of espresso test result for all the test files from sharedTest directory.

MaskedCarrot committed:
nit

MaskedCarrot committed:
lint fix

MaskedCarrot committed:
Merge branch 'add-resume-lesson-activity' into app-layer-mech-to-resume-checkpoints

MaskedCarrot committed:
nit

rt4914 reviewed (approved):
LGTM thanks.

MaskedCarrot committed:
Merge branch 'develop' into resume-exploration-in-domain-module

MaskedCarrot committed:
Merge branch 'resume-exploration-in-domain-module' into add-resume-lesson-activity

MaskedCarrot committed:
improved testing and nit fix

BenHenning commented on testing/src/main/java/org/oppia/android/testing/lightweightcheckpointing/ExplorationCheckpointTestHelper.kt:
```diff
@@ -24,19 +24,44 @@ import org.oppia.android.util.data.DataProviders.Companion.toLiveData
+/** The current exploration version of Fractions topic, story 0, exploration 0. */
+const val FRACTIONS_STORY_0_EXPLORATION_0_CURRENT_VERSION = 85
+
+/** The old exploration version of Fractions topic, story 0, exploration 0. */
```
```suggestion
/** An old, not up-to-date version of Fractions topic, story 0, exploration 0. */
```

BenHenning reviewed (commented):
Thanks @MaskedCarrot. Overall, PR LGTM other than a few nits--PTAL.

anandwana001 commented:
> nit thoughts.
> 
> Also, add a screenshot of espresso test result for all the test files from sharedTest directory.

Please add screenshot @MaskedCarrot

MaskedCarrot committed:
Merge branch 'resume-exploration-in-domain-module' into add-resume-lesson-activity

MaskedCarrot committed:
nit

MaskedCarrot committed:
Merge branch 'add-resume-lesson-activity' into app-layer-mech-to-resume-checkpoints

MaskedCarrot committed:
fixed test and nit

MaskedCarrot committed:
Merge remote-tracking branch 'origin/develop' into resume-exploration-in-domain-module

MaskedCarrot committed:
Merge branch 'resume-exploration-in-domain-module' into add-resume-lesson-activity

MaskedCarrot committed:
Merge branch 'add-resume-lesson-activity' into app-layer-mech-to-resume-checkpoints

MaskedCarrot committed:
nit

MaskedCarrot committed:
nit

MaskedCarrot committed:
nit

MaskedCarrot committed:
fix failing test

MaskedCarrot committed:
Merge branch 'resume-exploration-in-domain-module' into add-resume-lesson-activity

MaskedCarrot committed:
fix tests

MaskedCarrot committed:
Merge branch 'add-resume-lesson-activity' into app-layer-mech-to-resume-checkpoints

MaskedCarrot committed:
fixed tests

MaskedCarrot committed:
Merge branch 'add-resume-lesson-activity' into app-layer-mech-to-resume-checkpoints

MaskedCarrot committed:
added version to 2 more text proto

BenHenning reviewed (commented):
Thanks @MaskedCarrot. LGTM for codeowners.

Please assign back once based on develop & with passing CI.

anandwana001 reviewed (changes requested):
Test LGTM, but waiting for PR to get base branch update and test pass.

MaskedCarrot committed:
Merge remote-tracking branch 'origin/develop' into app-layer-mech-to-resume-checkpoints

anandwana001 reviewed (approved):
All tests are passing on CI, LGTM for code owners' files.

BenHenning reviewed (approved):
Thanks @MaskedCarrot. LGTM as well.

BenHenning commented:
Given all reviewers have approved & there are no outstanding conversations, merging.
